applications/CoSimulationApplication/python_scripts/solver_wrappers/external/flux_wrapper.py

```diff
@@ -0,0 +1,712 @@
+                    }
+            self.ImportCouplingInterface(interface_config)
+
+        # ---------------------------------------------------------------------
+        # Creation of the required DataImport instances for each Kratos exports
+        # ---------------------------------------------------------------------
+        # For each Kratos exported data
+        for export_data_name in self.settings["solver_wrapper_settings"]["export_data"].GetStringArray():
+            # Check additional info
+            self.settings["data"][export_data_name]["additional_info"].ValidateAndAssignDefaults(self.GetDefaultDataAdditionalInfoParameters())
+
+            # Get arguments from JSON file
+            model_part_name = self.settings["data"][export_data_name]["model_part_name"].GetString()
+            flux_quantity = self.settings["data"][export_data_name]["additional_info"]["flux_quantity"].GetString()
+            flux_region_list = self.settings["data"][export_data_name]["additional_info"]["flux_region_list"].GetStringArray()
+            initial_value = self.settings["data"][export_data_name]["additional_info"]["initial_value"].GetDouble()
+            variable_name = self.settings["data"][export_data_name]["variable_name"].GetString()
+            location = self.settings["data"][export_data_name]["location"].GetString()
+
+            # Check entities
+            self.CheckDataEntities (export_data_name)
+
+            # Create the Flux DataSupport instance
+            data_support = self.flux.CreateDataSupportOnRegion(model_part_name,flux_region_list,self.extruded_layers,self.full_device)
+
+            # Import nodes and elements of the model part in the Kratos Environnement
+            interface_config = {
+                    "model_part_name"  : model_part_name,
+                    "initial_value"    : initial_value,
+                    "variable_name"    : variable_name,
+                    "location"         : location
+                    }
+            self.ImportCouplingInterface(interface_config)
+
+            # Create the DataImport instance
+            self.flux.CreateDataImport (export_data_name,data_support,flux_quantity)
+
+        # Force first step initialize
+        super().Initialize()
+
+    def __del__(self):
+        """
+        Destructor for the FluxWrapper instance.
+
+        This method is responsible for cleaning up resources associated with the FluxWrapper instance before it is deleted.
+
+        Args:
+            self: An instance of the FluxWrapper class.
+
+        Note:
+            If the instance has a 'flux' attribute, this method closes the MultiPhysics session, saves and closes the Flux project.
+        """
+
+        if hasattr(self, 'flux') :
+            # Close the MultiPhysic session
+            self.flux.CloseMultiPhysics()
+            # Save and close the Flux project
+            self.flux.SaveProject()
+            self.flux.CloseProject()
```

Comment:
should this also be called in `Finalize`?

If so, then perhaps refactor to a separate fct

Reply:
Okay, but I don't know the details

Reply:
I don't think we should call it twice in finalize and also in del but let me ask the author.

Reply:
meant to call it in `Finalize` instead of `del`, as in my experience `del` is not reliable

but perhaps a flag would be useful, e.g. `flux_was_closed` which is set in `Finalize`. If for some reason the run crashes, and only `del` is called, then you could execute the closing of flux depending on this flag

Reply:
maybe you could also use [atexit](https://docs.python.org/3/library/atexit.html) instead of `del`